dnsmasq_interface.c

```diff
@@ -1200,77 +1200,72 @@ static void block_single_domain_regex(char *domain)
 
 	if(config.debug & DEBUG_QUERIES) logg("Added %s to cache", domain);
-
-	return;
 }
 
 int FTL_listsfile(char* filename, unsigned int index, FILE *f, int cache_size, struct crec **rhash, int hashsz)
 {
 	int name_count = cache_size;
 	int added = 0;
-	size_t size = 0;
-	char *buffer = NULL;
 	struct all_addr addr4 = {{{ 0 }}}, addr6 = {{{ 0 }}};
 	bool has_IPv4 = false, has_IPv6 = false;
+	unsigned char list = 0;
 
 	// Handle only gravity.list and black.list
 	// Skip all other files (they are interpreted in the usual format)
-	if(strcmp(filename, files.gravity) != 0 &&
-	   strcmp(filename, files.blacklist) != 0)
+	if(strcmp(filename, files.gravity) == 0)
+		list = GRAVITY_LIST;
+	else if(strcmp(filename, files.blacklist) == 0)
+		list = BLACK_LIST;
+	else
+		return cache_size;
+
+	// Jump to end of file to ensure dnsmasq does not
+	// try to read whatever might be in the file on
+	// disk when we return
+	fseek(f, 0, SEEK_END);
+
+	if(blockingstatus == BLOCKING_DISABLED)
+	{
+		logg("Skipping %s because blocking is disabled", filename);
 		return cache_size;
+	}
+	else
+	{
+		logg("Importing content of template %s from database", filename);
+
+	}
 
 	// Start timer for list analysis
 	timer_start(LISTS_TIMER);
 
+	// Start database interaction
+	if(!gravityDB_getTable(list))
+	{
+		logg("FTL_listsfile(%s, ...): Error getting table from database", filename);
+		return name_count;
+	}
+
 	// Get IPv4/v6 addresses for blocking depending on user configured blocking mode
 	prepare_blocking_mode(&addr4, &addr6, &has_IPv4, &has_IPv6);
 
 	// If we have neither a valid IPv4 nor a valid IPv6 but the user asked for
 	// blocking modes MODE_IP or MODE_IP_NODATA_AAAA then we cannot add any entries here
 	if(!has_IPv4 && !has_IPv6)
 	{
-		logg("ERROR: found neither a valid IPV4_ADDRESS nor IPV6_ADDRESS in setupVars.conf");
+		logg("ERROR: Cannot add domains from gravity because pihole-FTL found\n" \
+		     "       neither a valid IPV4_ADDRESS nor IPV6_ADDRESS in setupVars.conf");
 		return cache_size;
 	}
 
-	// Walk file line by line
-	bool firstline = true;
-	while(getline(&buffer, &size, f) != -1)
+	// Walk database table
+	const char *domain = NULL;
+	while((domain = gravityDB_getDomain()) != NULL)
 	{
-		char *domain = buffer;
-		// Skip hashed out lines
-		if(*domain == '#')
-			continue;
-
-		// Filter leading dots or spaces
-		while (*domain == '.' || *domain == ' ') domain++;
-
-		// Check for spaces or tabs
-		// If found, then this list is still in HOSTS format and we
-		// don't analyze it here. We only check the first line for
-		// efficiency reasons (strstr() is slow)
-		if(firstline &&
-		   (strstr(domain, " ") != NULL || strstr(domain, "\t") != NULL))
-		{
-			// Reset file pointer back to beginning of the list
-			rewind(f);
-			logg("File %s is in HOSTS format, please run pihole -g!", filename);
-			return name_count;
-		}
-		firstline = false;
-
 		// Skip empty lines
```

Comment:
This comment should be adjusted to "Skip empty domains", since there is no concept of "lines" in the database.